watson_developer_cloud/conversation_v1.py

```diff
@@ -11,108 +13,200 @@
+               application locally or outside of Bluemix. When running on
+               Bluemix, the credentials will be automatically loaded from the
+               `VCAP_SERVICES` environment variable.
+
+        :param str password: The password used to authenticate with the service.
+               Username and password credentials are only required to run your
+               application locally or outside of Bluemix. When running on
+               Bluemix, the credentials will be automatically loaded from the
+               `VCAP_SERVICES` environment variable.
+
+        """
+
+        WatsonService.__init__(
+            self,
+            vcap_services_name='conversation',
+            url=url,
+            username=username,
+            password=password,
+            use_vcap_services=True)
         self.version = version
 
     #########################
-    # counterexamples
+    # workspaces
     #########################
 
-    def create_counterexample(self, workspace_id, text):
+    def create_workspace(self,
+                         name=None,
+                         description=None,
+                         language=None,
+                         intents=None,
+                         entities=None,
+                         dialog_nodes=None,
+                         counterexamples=None,
+                         metadata=None,
+                         learning_opt_out=None):
         """
-        Create counterexample.
+        Create workspace.
 
-        Add a new counterexample to a workspace. Counterexamples are
-        examples that have been marked as irrelevant input.
+        Create a workspace based on component objects. You must provide workspace
+        components defining the content of the new workspace.
 
-        :param workspace_id: The workspace ID.
-        :param text: The text of a user input example.
+        :param str name: The name of the workspace.
+        :param str description: The description of the workspace.
+        :param str language: The language of the workspace.
+        :param list[CreateIntent] intents: An array of objects defining the intents for the workspace.
+        :param list[CreateEntity] entities: An array of objects defining the entities for the workspace.
+        :param list[CreateDialogNode] dialog_nodes: An array of objects defining the nodes in the workspace dialog.
+        :param list[CreateCounterexample] counterexamples: An array of objects defining input examples that have been marked as irrelevant input.
+        :param object metadata: Any metadata related to the workspace.
+        :param bool learning_opt_out: Whether training data from the workspace can be used by IBM for general service improvements. `true` indicates that workspace training data is not to be used.
+        :return: A `dict` containing the `Workspace` response.
+        :rtype: dict
         """
+        if intents is not None:
+            intents = [
```

Comment:
This is an implementation code readability issue, not an API issue, but can we extract this logic out so that it happens with a single call rather than seeing a bunch of lines of code for each parameter? We have similar logic in the base class right now for automatically converting None values, etc.

Reply:
Well, it only wraps to multiple lines because the 80 char line length restriction, but yes, we can look at creating a utility method for this.

Reply:
How about this:  Create this utility method:
```
def _convert_model(modelOrDict):
    if hasattr(x, "_to_dict"):
        return x._to_dict()
    return x
```
So that the code in the service can now look like:
```
        if intents is not None:
            intents = [ _convert_model(x) for x in intents ]
```

Reply:
You can make it better by detecting when it's an array and doing the for there

Reply:
@german I'm not sure I follow.  Are you saying that we should not assume that `intents` is an array -- as this code seems to do -- and instead we should wrap this code in a check that `intents` is an array?

Reply:
I was thinking that we can simplify the `_convert_model` code and make it work for all the cases. With some introspection we can see if is an array or not.